src/dbnode/integration/wide_query_test.go

```diff
@@ -0,0 +1,343 @@
+		blockSize     = time.Hour * 2
+		verifyTimeout = 2 * time.Minute
+	)
+
+	// Test setup
+	idxOpts := namespace.NewIndexOptions().
+		SetEnabled(true).
+		SetBlockSize(time.Hour * 2)
+	nsOpts := namespace.NewOptions().
+		SetIndexOptions(idxOpts).
+		SetRepairEnabled(false).
+		SetRetentionOptions(defaultIntegrationTestRetentionOpts)
+
+	nsID := testNamespaces[0]
+	nsMetadata, err := namespace.NewMetadata(nsID, nsOpts)
+	require.NoError(t, err)
+
+	// Set up file path prefix
+	postfix := atomic.AddUint64(&created, 1) - 1
+	filePathPrefix, err := ioutil.TempDir("", fmt.Sprintf("integration-test-%d", postfix))
+	require.NoError(t, err)
+
+	testOpts := NewTestOptions(t).
+		SetTickMinimumInterval(time.Second).
+		SetNamespaces([]namespace.Metadata{nsMetadata}).
+		SetFilePathPrefix(filePathPrefix)
+
+	require.NoError(t, err)
+	fsOpts := fs.NewOptions().SetFilePathPrefix(filePathPrefix)
+	decOpts := fsOpts.DecodingOptions().SetIndexEntryHasher(xtest.NewParsedIndexHasher(t))
+	fsOpts = fsOpts.SetDecodingOptions(decOpts)
+
+	testSetup, err := NewTestSetup(t, testOpts, fsOpts,
+		func(opt storage.Options) storage.Options {
+			return opt.SetWideBatchSize(batchSize)
+		})
+	require.NoError(t, err)
+	defer testSetup.Close()
+
+	// Start the server with filesystem bootstrapper
+	log := testSetup.StorageOpts().InstrumentOptions().Logger()
+	log.Debug("wide fetch test")
+	require.NoError(t, testSetup.StartServer())
+	log.Debug("server is now up")
+
+	// Stop the server
+	defer func() {
+		require.NoError(t, testSetup.StopServer())
+		log.Debug("server is now down")
+	}()
+
+	// Setup test data
+	now := testSetup.NowFn()()
+	indexWrites := make(TestIndexWrites, 0, seriesCount)
+	ids := make([]string, 0, seriesCount)
+	for i := 0; i < seriesCount; i++ {
+		// Keep in lex order.
+		padCount := i / 10
+		pad := ""
+		for i := 0; i < padCount; i++ {
+			pad = fmt.Sprintf("%so", pad)
+		}
+
+		id := fmt.Sprintf("foo%s-%d", pad, i)
+		ids = append(ids, id)
+		indexWrites = append(indexWrites, testIndexWrite{
+			id:    ident.StringID(id),
+			tags:  ident.MustNewTagStringsIterator("abc", fmt.Sprintf("def%d", i)),
+			ts:    now,
+			value: float64(i),
+		})
+	}
+
+	log.Debug("write test data")
+	client := testSetup.M3DBClient()
+	session, err := client.DefaultSession()
+	require.NoError(t, err)
+
+	start := time.Now()
+	indexWrites.Write(t, nsID, session)
+	log.Info("test data written", zap.Duration("took", time.Since(start)))
+
+	log.Info("waiting until data is indexed")
+	indexed := xclock.WaitUntil(func() bool {
+		numIndexed := indexWrites.NumIndexed(t, nsID, session)
+		return numIndexed == len(indexWrites)
+	}, verifyTimeout)
+	require.True(t, indexed)
+	log.Info("verified data is indexed", zap.Duration("took", time.Since(start)))
+
+	// Advance time to make sure all data are flushed. Because data
+	// are flushed to disk asynchronously, need to poll to check
+	// when data are written.
+	testSetup.SetNowFn(testSetup.NowFn()().Add(blockSize * 2))
```

Comment:
This feels a bit error prone. If some subcomponent took a copy of `testSetup.NowFn()`, it will not see this change.
Maybe provide your own `NowFn` function during construction of `testSetup` (I believe it is possible via `TestOptions.SetNowFn`) that captures a local time var, and then vary the value of that var when you need to advance the time.

Reply:
This is a fairly common pattern I was using from some of our existing integration tests, e.g. here:  https://github.com/m3db/m3/blob/master/src/dbnode/integration/disk_coldflush_test.go#L143

Reply:
Your point is valid but probably makes more sense in a general "integration test cleanup" task in future

Reply:
FWIW @linasm that is actually what happens under the hood, `testSetup.SetNowFn(...)` changes the copy that everyone already sees:
https://github.com/m3db/m3/blob/5bcb1ba/src/dbnode/integration/setup.go#L360:L364